File: frontend/src/lib/csv.ts
import type { InstructorEntry, Student } from '../types/app'
import { parseCsvText } from '../shared/csv/csvUtils'

function buildInstructorMap(instructors: InstructorEntry[]): Record<string, string> {
  const map: Record<string, string> = {}
  instructors.forEach(instructor => {
    const name = instructor.name.trim()
    if (!name) {
      return
    }
    instructor.codes
      .split(',')
      .map(code => code.trim())
      .filter(Boolean)
      .forEach(code => {
        map[code] = name
      })
  })
  return map
}

export function parseStudentsFromCsv(
  text: string,
  instructors: InstructorEntry[],
  fallbackDay: string
): Student[] {
  const rows = parseCsvText(text)
  const rawHeader = rows[0]?.join(',') ?? ''
  if (rows.length < 2) {
    return []
  }

  const courseMap = buildInstructorMap(instructors)
  let SERVICE_NAME = 0
  let CODE = 7
  let DAY = 4
  let TIME = 6
  let LOCATION = 15
  let SCHEDULE = 8
  let NAME = 18
  let PHONE = 20
  let series = true

  if (!rawHeader.includes('EventSchedule')) {
    SERVICE_NAME = 10
    CODE = 0
    DAY = 2
    TIME = 1
    LOCATION = 6
    SCHEDULE = 2
    NAME = 16
    PHONE = 17
    series = false
  }

  const students: Student[] = []

  for (let i = 1; i < rows.length; i += 1) {
    const row = rows[i]
    if (!row.length) {
      continue
    }

    const rawDay = row[DAY] ?? ''
    const dayValue =
      rawDay === 'Mo Tu We Th Fr' || rawDay === 'Mo,Tu,We,Th,Fr'
        ? 'Mo,Tu,We,Th,Fr'
        : rawDay || fallbackDay

    const code = (row[CODE] ?? '').trim()
    const serviceName = (row[SERVICE_NAME] ?? '').trim()
    const time = (row[TIME] ?? '').trim()
    const location = (row[LOCATION] ?? '').trim()
    const schedule = (row[SCHEDULE] ?? '').trim()
    const phone = (row[PHONE] ?? '').trim()

    let name = ''
    if (series) {
      const lastName = row[NAME] ?? ''
      const firstName = row[NAME + 1] ?? ''
      name = `${firstName} ${lastName}`.trim()
    } else {
      name = (row[NAME] ?? '').trim()
    }

    if (!code || !name) {
      continue
    }

    const instructor = courseMap[code] ?? ''
    const level = serviceName
    const id = `${code}-${name}-${time}-${dayValue}`.replace(/\s+/g, '-')

    students.push({
      id,
      service_name: serviceName,
      code,
      day: dayValue,
      time,
      location,
      schedule,
      name,
      phone,
      instructor,
      level,
    })
  }

  return students
}

File: backend/tasks/csv.go
package tasks

import (
	"fmt"
	"io"
	"strings"
)

type RosterStudent struct {
	Name       string `json:"name"`
	Phone      string `json:"phone"`
	Instructor string `json:"instructor"`
	Level      string `json:"level"`
}

type ClassRoster struct {
	Code        string          `json:"code"`
	ServiceName string          `json:"serviceName"`
	Day         string          `json:"day"`
	Time        string          `json:"time"`
	Location    string          `json:"location"`
	Schedule    string          `json:"schedule"`
	Instructor  string          `json:"instructor"`
	Students    []RosterStudent `json:"students"`
}

func ProcessCSVFromCSV(csvReader io.Reader, instructorMap map[string]string, fallbackDay string) ([]ClassRoster, int, error) {
	rows, err := readCSVRows(csvReader)
	if err != nil {
		return nil, 0, err
	}
	return ProcessCSVRows(rows, instructorMap, fallbackDay)
}

func ProcessCSVRows(rows []csvRow, instructorMap map[string]string, fallbackDay string) ([]ClassRoster, int, error) {
	if len(rows) == 0 {
		return nil, 0, fmt.Errorf("no rows to process")
	}

	classMap := map[string]*ClassRoster{}
	totalStudents := 0

	for _, row := range rows {
		serviceName := rowValue(row, "ServiceName", "Service", "Service Name")
		code := rowValue(row, "EventID", "Event Id", "ClassCode", "Code")
		day := rowValue(row, "DayOfTheWeek", "Day Of The Week", "DayOfWeek", "Day")
		timeValue := rowValue(row, "EventTime", "Time")
		location := rowValue(row, "Location", "Facility")
		schedule := rowValue(row, "EventSchedule", "Schedule")
		phone := rowValue(row, "AttendeePhone", "Phone")
		instructorFromRow := rowValue(row, "Instructor", "Instructor Name", "InstructorName", "Teacher", "Staff")

		name := rowValue(row, "AttendeeName", "Name")
		if strings.Contains(name, ",") {
			parts := strings.SplitN(name, ",", 2)
			name = strings.TrimSpace(parts[1]) + " " + strings.TrimSpace(parts[0])
		}
		if name == "" {
			firstName := rowValue(row, "FirstName", "First Name")
			lastName := rowValue(row, "LastName", "Last Name")
			if firstName != "" || lastName != "" {
				name = strings.TrimSpace(strings.Join([]string{firstName, lastName}, " "))
			}
		}

		if name == "" || code == "" {
			continue
		}

		if day == "" {
			day = fallbackDay
		}
		day = normalizeDay(day)
		if schedule == "" {
			schedule = day
		}

		instructor := strings.TrimSpace(instructorMap[code])
		if instructor == "" {
			instructor = strings.TrimSpace(instructorFromRow)
		}
		level := serviceName

		roster, ok := classMap[code]
		if !ok {
			roster = &ClassRoster{
				Code:        code,
				ServiceName: serviceName,
				Day:         day,
				Time:        timeValue,
				Location:    location,
				Schedule:    schedule,
				Instructor:  instructor,
				Students:    []RosterStudent{},
			}
			classMap[code] = roster
		} else if roster.Instructor == "" && instructor != "" {
			roster.Instructor = instructor
		}

		roster.Students = append(roster.Students, RosterStudent{
			Name:       name,
			Phone:      phone,
			Instructor: instructor,
			Level:      level,
		})
		totalStudents++
	}

	classes := make([]ClassRoster, 0, len(classMap))
	for _, roster := range classMap {
		classes = append(classes, *roster)
	}

	return classes, totalStudents, nil
}

func normalizeHeader(header string) string {
	clean := strings.TrimSpace(header)
	clean = strings.TrimPrefix(clean, "\uFEFF")
	clean = strings.TrimSpace(clean)
	clean = strings.ToLower(clean)
	return clean
}

func normalizeDay(day string) string {
	value := strings.TrimSpace(day)
	if value == "" {
		return value
	}
	if value == "Mo Tu We Th Fr" {
		return "Mo,Tu,We,Th,Fr"
	}
	normalized := strings.ToLower(value)
	switch normalized {
	case "monday":
		return "Mo"
	case "tuesday":
		return "Tu"
	case "wednesday":
		return "We"
	case "thursday":
		return "Th"
	case "friday":
		return "Fr"
	case "saturday":
		return "Sa"
	case "sunday":
		return "Su"
	}
	return value
}
